### PR TITLE
Tribal Section 3 Parsing & Validation

### LINKED ISSUE #1116: Tribal TANF Aggregate Data (03) File Parsing - As a data analyst, I want my data files to be parsed so that my submission can be validated/accepted by TDP
**Description:**
_Provide a brief background and justification for this issue_
Upon file upload, we expect TDP to attempt parsing per the spec outlined:
https://www.acf.hhs.gov/sites/default/files/documents/ofa/tanf_data_report_section3.pdf

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Allow Tribes to submit Closed Data based on the TANF data parsing spec
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 
- [ ] Ensuring file content type is tribal TANF. 
- [ ] Correctly formatted input data can be built into a Python data structure
- [ ] Completed Python data structure can be translated to database model(s)
- [ ] Incorrectly formatted data is identified and mistake(s) are logged.
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 
This is the same as TANF section 3. 

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Added support for parsing and validating Tribal TANF T6 records
Pull request closes #1116 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
- Submit both `ADS.E2J.FTP3.TS142` 
- Verify the file generates 3 Tribal T6 records.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Allow Tribes to submit aggregate Data based on the TANF data parsing spec
- [ ] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by raftmsohani — CHANGES_REQUESTED** [review]
Please add friendly_name to t6 fields

**Review by jtimpe — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Comment by elipe17** [comment]
@ADPennington this is up to date and is ready. Deploying for you now.

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=15,
+            endIndex=23,
+            required=True,
+            validators=[validators.isInLimits(0, 99999999)],
+        ),
+        Field(
+            item="5B",
+            name="NUM_APPROVED",
+            friendly_name="total number of approved cases",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of approved applications",
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=39,
+            endIndex=47,
+            required=True,
+            validators=[validators.isInLimits(0, 99999999)],
+        ),
+        Field(
+            item="6B",
+            name="NUM_DENIED",
+            friendly_name="total number of denied",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of denied applications",
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=131,
+            endIndex=139,
+            required=True,
+            validators=[validators.isInLimits(0, 99999999)],
+        ),
+        Field(
+            item="9C",
+            name="NUM_2_PARENTS",
+            friendly_name="total number of two parents families",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of two parent families",
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=171,
+            endIndex=179,
+            required=True,
+            validators=[validators.isInLimits(0, 99999999)],
+        ),
+        Field(
+            item="11B",
+            name="NUM_NO_PARENTS",
+            friendly_name="total number of no-parent family",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of no parent family",
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=155,
+            endIndex=163,
+            required=True,
+            validators=[validators.isInLimits(0, 99999999)],
+        ),
+        Field(
+            item="10C",
+            name="NUM_1_PARENTS",
+            friendly_name="total number of one parent families",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of one-parent families",
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=91,
+            endIndex=103,
+            required=True,
+            validators=[validators.isInLimits(0, 999999999999)],
+        ),
+        Field(
+            item="8B",
+            name="NUM_FAMILIES",
+            friendly_name="total of number of families",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of families",
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=79,
+            endIndex=91,
+            required=True,
+            validators=[validators.isInLimits(0, 999999999999)],
+        ),
+        Field(
+            item="8A",
+            name="NUM_FAMILIES",
+            friendly_name="number of families",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of families",
```

**Review by ADPennington — CHANGES_REQUESTED** [review]
looks good @elipe17. a few suggestions on inconsistencies in the friendly names. I imagine these are also incorrect in [TANF T6 schema](https://github.com/raft-tech/TANF-app/blob/develop/tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py) as well, so a follow-up ticket forthcoming.

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=171,
+            endIndex=179,
+            required=True,
+            validators=[validators.isInLimits(0, 99999999)],
+        ),
+        Field(
+            item="11B",
+            name="NUM_NO_PARENTS",
+            friendly_name="total number of no parent family",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of no parent families",
```

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+            type="number",
+            startIndex=131,
+            endIndex=139,
+            required=True,
+            validators=[validators.isInLimits(0, 99999999)],
+        ),
+        Field(
+            item="9C",
+            name="NUM_2_PARENTS",
+            friendly_name="total number of two parent families",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t6.py]:
```suggestion
            friendly_name="total number of two-parent families",
```

**Review by ADPennington — COMMENTED** [review]
caught a couple more @elipe17. I can approve upon commit.

**Comment by elipe17** [comment]
> caught a couple more @elipe17. I can approve upon commit.

Updated!

**Review by ADPennington — APPROVED** [review]
🚀

**Comment by andrew-jameson** [comment]
@elipe17 Can you resolve merge conflicts here?

**Comment by elipe17** [comment]
> @elipe17 Can you resolve merge conflicts here?

@andrew-jameson resolved :)

**Pull request merged**